pymemcache/client/retrying.py

```diff
@@ -0,0 +1,183 @@
+        attempts=2,
+        retry_delay=0,
+        retry_for=None,
+        do_not_retry_for=None
+    ):
+        """
+        Constructor for RetryingClient.
+
+        Args:
+          client: Client|PooledClient|HashClient, inner client to use for
+            performing actual work.
+          attempts: optional int, how many times to attempt an action before
+            failing. Must be 1 or above. Defaults to 2.
+          retry_delay: optional int|float, how many seconds to sleep between
+            each attempt.
+            Defaults to 0.
+
+          retry_for: optional tuple[class]|Iterable[class], what exceptions to
+            allow retries for. Will allow retries for all exceptions if None.
+            Example:
+                `(MemcacheClientError, MemcacheUnexpectedCloseError)`
+            Accepts any class that is a subclass of Exception.
+            Defaults to None.
+
+          do_not_retry_for: optional None|tuple[class]|Iterable[class], what
+            exceptions should be retried. Will not block retries for any
+            Exception if None.
+            Example:
+                `(IOError, MemcacheIllegalInputError)`
+            Accepts any class that is a subclass of Exception.
+            Defaults to None.
+
+        Exceptions:
+          ValueError: If `attempts` is not 1 or above.
+          ValueError: If `retry_for` or `do_not_retry_for` is not None, tuple or
+            Iterable.
+          ValueError: If any of the elements of `retry_for` or
+            `do_not_retry_for` is not a subclass of Exception.
+          ValueError: If there is any overlap between `retry_for` and
+            `do_not_retry_for`.
+        """
+
+        if attempts < 1:
+            raise ValueError("`attempts` argument must be at least 1.")
+
+        self._client = client
+        self._attempts = attempts
+        self._retry_delay = retry_delay
+        self._retry_for = _ensure_tuple_argument("retry_for", retry_for)
+        self._do_not_retry_for = _ensure_tuple_argument(
+            "do_not_retry_for", do_not_retry_for
+        )
+
+        # Verify no overlap in the go/no-go exception collections.
+        for exc_class in self._retry_for:
+            if exc_class in self._do_not_retry_for:
+                raise ValueError(
+                    "%s was present in both `retry_for` and "
+                    "`do_not_retry_for`." % repr(exc_class)
+                )
+
+        # Take dir from the client to speed up future checks.
+        self._client_dir = dir(self._client)
+
+    def _retry(self, name, func, *args, **kwargs):
+        """
+        Workhorse function, handles retry logic.
+
+        Args:
+          name: str, Name of the function called.
+          func: callable, the function to retry.
+          *args: args, array arguments to pass to the function.
+          **kwargs: kwargs, keyword arguments to pass to the function.
+        """
+        for attempt in range(1, self._attempts + 1):
```

Comment:
Can't we start our range from `0` to `self._attempts` to avoid to bump `self._attempts`?

By example if user pass `attempts=1` that will avoid a mental gymnastic to get our starting point during code reading. You protected the execution from values inferior to `1`, so I don't think that we need to tweak it during runtime.

Reply:
Sure. I'm am not opinionated on it.
https://github.com/martinnj/pymemcache/commit/b0782c595dafa92c1608abd8f749c255ebf71edd